Clarify client behavior when HRR acceptance signal is malformed (#470)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

The payload of the extension has the following structure:

~~~~
    enum { outer(0), inner(1) } ECHClientHelloType;

    struct {

>        ECHClientHelloType type;

Reviewer: Does this field need to be explicit?  Empty/not is a pretty good discriminator.

Author: This was a change made in https://github.com/tlswg/draft-ietf-tls-esni/pull/422. The reasoning is that 8446 requires extensions to be solicited by the client, meaning in order for there to be an ECH extension in HRR, there must be an ECH extension in CHInner.

Reviewer: Yes, but the extension can be empty and achieve the same end.

Reviewer 2: Yeah, whether or not we need the octet indicator was raised in a separate PR. I suggest we move this off to a separate issue if we want to change it.

Reviewer 3: Agreed that empty or not is a fine discriminator, though I'm not sure how to encode that cleanly in the TLS presentation language.

Reviewer: We switch on external variables in other contexts.  Continuing to think of the presentation language as a formal grammar is probably not helpful..

decreasing order of preference, to be used by the client in subsequent
connection attempts. These are known as the server's "retry configurations".

Finally, when the client offers the "encrypted_client_hello", if the payload is

> the `inner` variant and the server responds with HelloRetryReqwuest, it MUST

Author: I'm slightly worried that the "MUST" here violates extensions guidelines.

Reviewer 2: How is this different from, say, requiring the server to send a key_share extension in SH?

Author: Not any different, I don't think.

> include an "encrypted_client_hello" with the following payload:

Reviewer: ```suggestion
include an "encrypted_client_hello" extension with the following payload:
```

~~~
    struct {

>        opaque confirmation [8];

Reviewer 2: ```suggestion
       opaque confirmation[8];
```

>     } ECHHelloRetryRequest;

Reviewer: Why isn't this in a `select {}` block for the extension?

Author: [We had a thread on this, but I accidentally resolve it and I can no longer find it. So ist das Leben.] We're following the precedent set by the "key_share" extension https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8.

### Handling HelloRetryRequest {#client-hrr}

When the server sends a HelloRetryRequest, the client checks for the presence of
an "encrypted_client_hello" extension. If none is found, then the client
presumes rejection and handles the HelloRetryRequest using ClientHelloOuter.
(Note that the client-facing server does not send this extension in its
HelloRetryRequest.) Otherwise it proceeds as follows.

If the extension's payload has a length other than 8, then the client aborts the

> handshake with an "illegal_parameter" alert. If the payload length is equal to

Reviewer 3: How about we capture this in syntax, as `opaque HelloRetryRequestECH[8]` or whatever? I think syntax errors are already implicitly fatal and come with an alert (`decode_error`). Less special-case bits.

Looking at RFC8446, we seem to have a couple styles here. Sometimes we just define separate types for each context and link them in prose: `KeyShareClientHello`, `KeyShareHelloRetryRequest`, `KeyShareServerHello`. Sometimes we define one structure with a select:

```
      struct {
          select (Handshake.msg_type) {
              case new_session_ticket:   uint32 max_early_data_size;
              case client_hello:         Empty;
              case encrypted_extensions: Empty;
          };
      } EarlyDataIndication;

      struct {
          select (Handshake.msg_type) {
              case client_hello:
                   ProtocolVersion versions<2..254>;

              case server_hello: /* and HelloRetryRequest */
                   ProtocolVersion selected_version;
          };
      } SupportedVersions;
```

I guess the latter is a little interesting because, for HelloRetryRequest, `Handshake.msg_type` is still `server_hello`. Maybe that's why we did key shares in prose...

(Interestingly, our naming scheme for ECH is funny because it doesn't actually tell you the message, only who sends it. We say ServerECH rather than ECHEncryptedExtensions. Maybe time to rename since the server sends two of them now...)

Reviewer 3: > I guess the latter is a little interesting because, for HelloRetryRequest, Handshake.msg_type is still server_hello. Maybe that's why we did key shares in prose...

Ah yes, that's exactly why key share doesn't do this. So I guess the single struct with `select` is the convention, but if you need to distinguish SH and HRR, as we do, do it in prose:
https://github.com/tlswg/tls13-spec/commit/9a9c3295140bf4880ca031a6512e350da108713b

Author: Something like this?
```
      struct {
          select (Handshake.msg_type) {
              case server_hello: // Well, HRR, but this isn't a codepoint AFAIK
                   opaque hrr_ech_accept_confirmation [8];
              case encrypted_extensions:
                   ECHConfigList retry_configs;
          };
      } ServerECH;
```

Reviewer 3: Using a mix of the two styles seems a little odd. Normally, when extensions use a `select (Handshake.msg_type)`, we put everything in it, so the ClientHello extension would go here too. Though I suppose we don't have a SH extension so HRR using the SH code point does work. But I was thinking we just go the key_share route and define three structures:

```
    struct {
       ClientECHType type;
       select (ClientECH.type) {
           case outer:
               HpkeSymmetricCipherSuite cipher_suite;
               uint8 config_id;
               opaque enc<0..2^16-1>;
               opaque payload<1..2^16-1>;
           case inner:
               Empty;
       };
    } ECHClientHello;

    struct {
       opaque confirmation[8];
    } ECHHelloRetryRequest;
    // Or, equivalently, opaque ECHHelloRetryRequest[8], but the struct might be easier to read.

    struct {
       ECHConfigList retry_configs;
    } ECHEncryptedExtensions;
```

Renaming things might be too much churn, I dunno. That's just the convention RFC8446 used for key share.

Author: This looks good to me. @chris-wood, what do you think?

Reviewer 2: Seems reasonable to me!

Author: Done.

Reviewer 2: ```suggestion
handshake with an "decode_error" alert. If the payload length is equal to
```

Reviewer 2: (@cjpatton did you accidentally resolve this?)

8, then the client checks if the payload is equal to `hrr_accept_confirmation`
as defined in {{backend-server}}. If so, then it presumes acceptance and handles
handles the HelloRetryRequest using ClientHelloInner. Otherwise, it presumes
rejection.

[[OPEN ISSUE: Depending on what we do for issue#450, it may be appropriate to
change the client behavior if the HRR payload is missing or malformed.]]

The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
the second ClientHelloOuter for any referenced extensions. It then encrypts
the new EncodedClientHelloInner value as a second message with the previous
HPKE context as described in {{encrypting-clienthello}}.

> ## GREASE ECH {#grease-ech}

Reviewer 3: If we allow the server to send HRR.ech with invalid value as rejection, does that mean that, by sending GREASE ECH, the client is obligated to handle this case? Do we need some text here? (Like the client checks HRR.ECH syntactically but interprets all values as rejection.)

Author: Yeah, I think the client is obligated to handle this case. In particular, it should parse the extension. (Isn't the server obligated to do the same for the GREASE ECH extension?)

Reviewer 3: Right. What I'm saying is the server specification doesn't distinguish GREASE ECH from normal ECH that fails to decrypt (and one hopes it cannot!). But the client specification has a big branch between GREASE ECH and normal ECH, so we probably need text here if we want that to work. (And we need a big branch because a GREASE ECH client cannot actually run the steps to compute the confirmation. There's no ClientHelloInner or anything.)

Author: Do you think there needs to be text around this in this PR? I would think it would make sense to address this once we have text for #450.

Reviewer 2: It seems somewhat separable. We can just say that, if the client send GREASE ECH, then it should always assume rejection and use CHOuter.

accepted by checking the message for an "encrypted_client_hello" extension with
an 8-byte payload equal to `hrr_accept_confirmation` as defined in
{{backend-server}}. If found, the client presumes acceptance and handles the
HelloRetryRequest using ClientHelloInner. Otherwise, if the extension is missing
or its payload is not equal to `hrr_accept_confirmation`, the client presumes

> rejection and handles the HelloRetryRequest using ClientHelloOuter. Note that

Reviewer 2: What if the payload is present but malformed, e.g., 1 byte long? Would we want the client to abort then? (Failing early and loudly)

Author: A 1-byte *can't* equal `hrr_accept_confirmation`, so the client would presume rejection.

Reviewer 2: Right, but peers typically abort when values are malformed or otherwise incorrect. Presuming rejection in this case seems to break with that pattern.

Author: Offline we agreed that the client should abort if the payload is malformed.
